pbkdf2/src/simple.rs

```diff
@@ -69,65 +70,26 @@ pub fn pbkdf2_simple(password: &str, c: u32) -> io::Result<String> {
 /// * `password` - The password to process
 /// * `hashed_value` - A string representing a hashed password returned by
 /// `pbkdf2_simple`
-pub fn pbkdf2_check(password: &str, hashed_value: &str)
-    -> Result<(), CheckError> {
-    let mut iter = hashed_value.split('$');
-
-    // Check that there are no characters before the first "$"
-    if iter.next() != Some("") { Err(CheckError::InvalidFormat)?; }
-
-    // Check the name
-    if iter.next() != Some("rpbkdf2") { Err(CheckError::InvalidFormat)?; }
-
-    // Parse format - currenlty only version 0 is supported
-    match iter.next() {
-        Some(fstr) => {
-            match fstr {
-                "0" => { }
-                _ => return Err(CheckError::InvalidFormat)
-            }
-        }
-        None => return Err(CheckError::InvalidFormat)
-    }
-
-    // Parse the iteration count
-    let c = match iter.next() {
-        Some(pstr) => match base64::decode(pstr) {
-            Ok(pvec) => {
-                if pvec.len() != 4 { return Err(CheckError::InvalidFormat); }
-                BigEndian::read_u32(&pvec[..])
-            }
-            Err(_) => return Err(CheckError::InvalidFormat)
-        },
-        None => return Err(CheckError::InvalidFormat)
-    };
-
-    // Salt
-    let salt = match iter.next() {
-        Some(sstr) => match base64::decode(sstr) {
-            Ok(salt) => salt,
-            Err(_) => return Err(CheckError::InvalidFormat)
-        },
-        None => return Err(CheckError::InvalidFormat)
+pub fn pbkdf2_check(password: &str, hashed_value: &str) -> Result<(), CheckError> {
+    let parts: Vec<_> = hashed_value.split('$').collect();
+
+    // check the format of the input: there may be no tokens before the first
+    // and after the last `$`, tokens must have correct information and length.
+    let (count, salt, hash) = match parts[..] {
```

Comment:
I think you can remove the allocation by using:
```rust
let mut p = hashed_value.split('$');
let buf: [Option<&str>; 8] = [
    p.next(), p.next(), p.next(), p.next(),
    p.next(), p.next(), p.next(), p.next(),
];
let (count, salt, hash) = match buf {
    [
        Some(""), Some("rpbkdf2"), Some("0"), Some(count),
        Some(salt), Some(hash), Some(""), None,
    ] => (count, salt, hash),
    _ => return Err(CheckError::InvalidFormat),
};
```

Also it looks like `Err(CheckError::InvalidFormat)?` is discouraged (e.g. by clippy), so we probably better to replace it with an explicit `return`.

Reply:
Thank you for your suggestion. I will remove the unnecessary allocation (I wasn't happy with it, too) by applying your changes :smile: 
Unfortunately, I don't think, that Rust 1.22 will support this code as well.